### ISSUE TITLE
Normalize image encoders interface to the text encoder interface

### ISSUE THREAD

**Issue by KennethEnevoldsen** [embeddings-benchmark/mteb#1606]
As discussed in this [PR](https://github.com/embeddings-benchmark/mteb/pull/1551) we would like to normalize the interfaces to allow for:

```py
model.encode(sentences)
model.encode(images)
model.encode(documents) # can be image and text
```

**Comment by isaac-chung** [comment]
Adding to [this comment](https://github.com/embeddings-benchmark/mteb/pull/1551#discussion_r1876142413), the encoder can then be something like:

```python
class MultiModalEncoder:
    """For MIEB, MOEB, etc.
    """

    def __init__(self, device: str | None,**kwargs: Any):
        pass

    def encode(
        self,
        inputs: Sequence[str] | Sequence[Image.Image] | DataLoader | Sequence[Any],
        *,
        task_name: str,
        prompt_type: PromptType | None = None,
        **kwargs: Any,
    ) -> np.ndarray:
        """
        inputs: Handles uni-modal or multi-modal inputs. For example:
            - text only: Sequence[str]
            - image and text: Sequence[Any] -> Sequence[tuple[Image.Image, str]]
            - image dataloader and text: Sequence[Any] -> tuple[DataLoader, list[str]]
            - images and text: Sequence[Any] -> Sequence[tuple[list[Image.Image], str]]
        
        This can potentially wrap existing get_fused_embedding methods as well.
        """
        pass
```

**Comment by isaac-chung** [comment]
This may be applicable to audio + video as well. Tagging @Muennighoff too for MOEB.

**Comment by KennethEnevoldsen** [comment]
Hmm maybe:
```
inputs: Iterable[mteb.document] | Iterable[str] | Iterable[image] | Iterable[audio]

"""
        inputs: Handles uni-modal or multi-modal inputs. For example:
            - text only: Iterable[str]
            - images: Iterable[image]
            - ...
            - Multimodal: Iterable[mteb.document]. A document is a mteb object that contain text, image, audio. It can even contain text metadata such as title. These attributes can be acceced using document.body, document.title, document.images, document.audio. You can also get the relative position of the modalities in the text using document.text which is formatting as follows: "We can see in the picture <image-1>", where "<image-1>" refers to document.images[1].
"""
```

Currently documents are not needed for anything in MTEB, but it would simplify a lot of cases (retrieval + joint image + text cases). We don't need to add the ".text" attribute to start with.

@isaac-chung wdyt?

**Comment by Samoed** [comment]
Maybe we could pass dataloaders to the `encode` function, with each dataloader handling task-specific logic?

**Comment by KennethEnevoldsen** [comment]
Isn't a dataloader an Iterable?

**Comment by Samoed** [comment]
Yes, it's iterable, but we can specify for it interface with some methods

**Comment by isaac-chung** [comment]
Thanks! Based on @tomaarsen 's [comment here](https://github.com/UKPLab/sentence-transformers/issues/3118#issuecomment-2520870305), I was leaning away from a Document class, especially when we have user feedback about big changes to the interface. But I also understand that it's a clearer way to define which modalities are present.

Would it be possible to merge your suggestion with part of [this proposal](https://github.com/embeddings-benchmark/mteb/issues/1606#issuecomment-2567501893)? Multimodal inputs can still be accepted in tuples, in which the input types are sorted in alphabetical order (i.e. audio, image, text, video, heat map, depth map, etc). 

Inputs can still be directly passed in without mandatory casting inputs into Documents. That said, we could say that we'd prefer Documents moving forward (by placing Documents first as shown now and/or in the docs)?

Would like to see how DataLoader fits in here as well.

**Comment by isaac-chung** [comment]
Yeah I think that's almost all the cases! With audio, it could look something like this:

```python
# single modalities 
text: Iterable[str]
images: Iterable[Image.Image]
audio: Iterable[audio]

# current mixed cases from MIEB 
image_and_text: Iterable[tuple[Image.Image, str]]
image_dataloader_and_text: Iterable[tuple[DataLoader, list[str]]]
images_and_text: Iterable[tuple[list[Image.Image], str]]

# potential MAEB mixed case
audio_and_text: Iterable[tuple[audio,text]

mixed: image_and_text | image_dataloader_and_text | images_and_text | audio_and_text

# All together 
inputs: text | images | audio | mixed
```

**Comment by Samoed** [comment]
I think we can do like this:
```python
class MultiModalEncoder:
    """
    For MTEB, MIEB, MOEB, etc.
    """

    def __init__(self, device: str | None,**kwargs: Any):
        pass

    def encode_text(inputs, ...) -> np.ndarray:
         ...

    def encode_image(inputs, ...) -> np.ndarray:
         ...

    def encode_image_and_text(inputs, ...) -> np.ndarray:
         ...

    def encode(
        self,
        inputs: Sequence[str] | Sequence[Image.Image] | DataLoader | Sequence[Any],
        *,
        task_name: str,
        prompt_type: PromptType | None = None,
        **kwargs: Any,
    ) -> np.ndarray:
        """
        inputs: Handles uni-modal or multi-modal inputs. For example:
            - text only: Sequence[str]
            - image and text: Sequence[Any] -> Sequence[tuple[Image.Image, str]]
            - image dataloader and text: Sequence[Any] -> tuple[DataLoader, list[str]]
            - images and text: Sequence[Any] -> Sequence[tuple[list[Image.Image], str]]
        
        This can potentially wrap existing get_fused_embedding methods as well.
        """
        if task.modalities == ["text"]:
             return self.encode_text(inputs, ...)
       elif task.modalities == ["image"]:
            return self.encode_image(inputs, ...)
       elif task.modalities == ["text", "image"]:
            return self.encode_image_and_text(inputs, ...)  # currenly get_fused_embeddings
        pass
```

**Comment by KennethEnevoldsen** [comment]
Hmm understand that we in practice might implement `encode_image_and_text`, but I am not sure I would want in the in the interface (simply leave it to the model to wrap the `def encode`.

However it seems like we are almost settled on an `encode`. It seems like the primary issue is:

```
tuple[DataLoader, list[str]]
# eq. to 
tuple[Iterable[image], Iterable[str]]
```

This makes it easier to split up images and text and encode them separately than the

```
Iterable[Multimodal]
```

Though the second one allows for multiple images pr. text, no images, fully multimodal.

As I understand it no task uses this interface (but we could easily imagine one, e.g. wikipedia retrieval where the article contains images). So we could consider simply leaving it out for now (`Sequence[Any]`)


I must say that I favor the second interface It would also simplify the interface quite notably to now allow multiple iterables. However I also see that it is not in line with what we currently have for MIEB and it can be practical to have modalities split up.

A solution that you could do it:

```
batch_size = 100

images, texts = [], []
for i in range(100):
  text, images = next(dataloader)

images_emd = model._encode_images(images)
text_end = model._encode_text(texts)

# fuse
# repeat until dataloader is empty)
```

I think this is partly what @Samoed is suggesting. With the specialized function as shorthands (if implemented we simply call those)

**Comment by Samoed** [comment]
Also we can implement Document class which could also handle Image/Audio, but I think this is too complex

**Comment by KennethEnevoldsen** [comment]
> Also we can implement Document class which could also handle Image/Audio, but I think this is too complex

For now we can leave it out (no need to add it unless we have a task that needs it)

**Comment by isaac-chung** [comment]
> It seems like the primary issue is: 
tuple[DataLoader, list[str]]
 eq. to 
tuple[Iterable[image], Iterable[str]]
> This makes it easier to split up images and text and encode them separately than the
Iterable[Multimodal]
> Though the second one allows for multiple images pr. text, no images, fully multimodal.

Isn't the first one also multimodal? Multiple images can be yielded per each text as needed (handled in the implementation, not the interface). 

Bottom line is that mteb is based on huggingface datasets, so I think it's easier to discuss this with examples. So far, I believe we can already treat each sample as a "Document". In [this dataset](https://huggingface.co/datasets/fusing/fill50k), each sample has 2 images and 1 text. If the dataset is streamed, one row is returned with 2 images and 1 text, and can be specified by col name. `Sequence[Any]` would work here, where `Any` is `tuple[Image.Image, Image.Image, str]`. 

We can also load `ImageTextDataset` into a DataLoader (in ImageTextPairClassification), where the DataLoader yields batches of image/text pairs. Perhaps the DataLoader, along with a custom "Dataset", is a current example of the 2nd option?

**Comment by Samoed** [comment]
Datasets are integrated with torch.utils.data.Dataloader, maybe we don't need dataloader at all https://huggingface.co/docs/datasets/use_with_pytorch#data-loading

**Comment by isaac-chung** [comment]
> Datasets are integrated with torch.utils.data.Dataloader, maybe we don't need custom dataloader at all https://huggingface.co/docs/datasets/use_with_pytorch#data-loading

Maybe this could be a refactor opportunity?

**Comment by KennethEnevoldsen** [comment]
I think hf_subset and split are easy adds.

re. interface I am not sure I understand the bytes in the above?

I generally think this is way too complicated 
```
inputs: Sequence[str] | Sequence[EncodeInput] | Sequence[Image.Image] |Sequence[bytes] | DataLoader, # bytes for audio
```

but I like the dict approach. We could simply hand of the 

```
encode(
...
inputs: IterableDataset
...
):
```
Where Iterable dataset is a dict[str, Iterable[Image] |Sequence[Audio] | Sequence[text]]
with the matching key 
```
{"text": [...], "image": [...], "audio": [...]} 
```

we can potentially add optional keys like
```
# for retrieval
{
"text": [...],  # default text
"image": [...], 
"audio": [...], 
"body: [...], # models are allowed to construct the text from the body + title if they wish
"title: [...],
}
```

This works perfectly fine and we could easily adapt existing models:
```py
from datasets import load_dataset

import torch
ds = load_dataset('DDSC/lcc', split='train', streaming=True) # small ds
dataloader = torch.utils.data.DataLoader(ds, num_workers=0, batch_size=3)
sample = next(iter(dataloader))
sample
# {'text': ['Og det kan blive svært, hvis udlændinge både skal regne med at betale den høje, danske skat og samtidig at få ekstraudgifter til lægebesøg eller skolegang for deres børn.',
#   'Antalya, Tyrkiet - Billig leje af bil I samarbejde med Cartrawler kan vi herunder tilbyde dig den billigste pris på Antalya, Tyrkiet - Billig leje af bil.',
#   '" 40 børn løb og leg med Øens hold-karavanen, ved den sidste træning før sommerferien den 22. juni 2011.'],
#  'label': ['negativ', 'positiv', 'neutral']}
```


It does not handle >1 image pr. text. Any ideas/thoughts?

**Comment by Samoed** [comment]
I've added bytes for audio since I'm not entirely sure about their format, but it seems that `Audio` from `datasets` is used in `MAEB`.  

Also, @tomaarsen, could you take a look at the proposed method? It would be great if new models used the same keys for inputs.

**Comment by Samoed** [comment]
Maybe we could modify our tasks to output `dataloaders` instead of `datasets`, allowing the use of custom collators. However, some models might require their own transformations, which could be handled in the `encode` function.
